src/funnel.test.ts

```diff
@@ -0,0 +1,1049 @@
+    const foo = funnel(mockFn, {
+      reducer: constant(undefined),
+      invokedAt: "end",
+      burstCoolDownMs: UT,
+    });
+    foo.call();
+    await sleep(2 * UT);
+
+    expect(mockFn).toHaveBeenCalledTimes(0);
+  });
+
+  it("supports multiple arguments", () => {
+    const mockFn = vi.fn();
+    const foo = funnel(mockFn, {
+      reducer: (ret: unknown, a: number, b: string, c: boolean) => [
+        ret,
+        a,
+        b,
+        c,
+      ],
+      invokedAt: "start",
+      burstCoolDownMs: UT,
+    });
+    foo.call(1, "a", true);
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith([undefined, 1, "a", true]);
+  });
+
+  test("reducer isn't called again when 'invokedAt: start'", async () => {
+    const mockFn = vi.fn(ARGS_COLLECTOR);
+    const foo = funnel(doNothing(), {
+      reducer: mockFn,
+      invokedAt: "start",
+      burstCoolDownMs: UT,
+    });
+    foo.call("a");
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith(undefined, "a");
+
+    foo.call("b");
+    foo.call("c");
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+
+    await sleep(2 * UT);
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe("non-trivial (>0ms) timer duration", () => {
+  describe("delay timer", () => {
+    test("invokedAt: start", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "start",
+        delayMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+
+      foo.call("d");
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      // "b" and "c" have been ignored because the funnel is only configured to
+      // run the function at the start of the delay window, meaning anything
+      // after that is simply dropped.
+      expect(mockFn).toHaveBeenLastCalledWith(["d"]);
+    });
+
+    test("invokedAt: both", async () => {
+      const mockFn = vi.fn();
+      const foo = funnel(mockFn, {
+        reducer: ARGS_COLLECTOR,
+        invokedAt: "both",
+        delayMs: UT,
+      });
+      foo.call("a");
+      foo.call("b");
+      foo.call("c");
+
+      expect(mockFn).toHaveBeenCalledTimes(1);
+      expect(mockFn).toHaveBeenLastCalledWith(["a"]);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+      expect(mockFn).toHaveBeenLastCalledWith(["b", "c"]);
+
+      foo.call("d");
+      foo.call("e");
+      foo.call("f");
+
+      expect(mockFn).toHaveBeenCalledTimes(2);
+
+      await sleep(UT);
+
+      expect(mockFn).toHaveBeenCalledTimes(3);
+      expect(mockFn).toHaveBeenLastCalledWith(["d", "e", "f"]);
+    });
+
+    it("never fires when only delay is used and invoked only at end", async () => {
```

Comment:
**question:** is this behavior desirable?

i know it's technically to-spec, but i think it's counterintuitive. i think we should disallow this at the type-level

Reply:
Yea, that's a good idea to prevent this at the type level. My only concern is that the typing might constraint certain cases where people might build the timing policy object dynamically, e.g.:

```ts
const options = {};

if (...) {
  options.invokedAt = "end";
} else {
  options.invokedAt = "start";
}

if (...) {
  options.delayMs = 500;
} else {
  options.burstCoolDownMs = 500;
}

const x = funnel(..., options);
```

Albeit this is a contrived example, and perhaps preventing accidental bad edge-cases is more valuable.
```

Reply:
i can't really think of a case where the timing object is built dynamically. i would suggest we prevent this now, and if we decide to change this in the future we can always make the typing less strict, which is easier than the reverse.